src/metrics/instance_state_per.go

```diff
@@ -150,6 +154,59 @@ var ankaInstanceStatePerMetrics = []InstanceStatePerMetric{
 			}
 		},
 	},
+	{
+		BaseAnkaMetric: BaseAnkaMetric{
+			metric: CreateGaugeMetricVec("anka_instance_max_age_per_template_seconds", "Age of oldest Instance in a particular state, per Template. Visible only for templates with at least one instance (label: state, template_uuid, template_name)", []string{"state", "template_uuid", "template_name"}),
+			event:  events.EVENT_VM_DATA_UPDATED,
+		},
+		HandleData: func(instances []types.Instance, metric *prometheus.GaugeVec) {
+			var InstanceAgePerTemplateMaximumMap = map[string]map[string]int{}
+			var instanceTemplates []string
+			var instanceTemplatesMap = map[string]string{}
+			now := time.Now()
+			for _, instance := range instances {
+				instanceTemplates = append(instanceTemplates, instance.Vm.TemplateUUID)
+				instanceTemplatesMap[instance.Vm.TemplateUUID] = instance.Vm.TemplateName
+			}
+			instanceTemplates = uniqueThisStringArray(instanceTemplates)
+			for _, wantedState := range types.InstanceStates {
+				if _, ok := InstanceAgePerTemplateMaximumMap[wantedState]; !ok {
+					InstanceAgePerTemplateMaximumMap[wantedState] = make(map[string]int)
+				}
+				for _, wantedInstanceTemplate := range instanceTemplates {
+					age := 0.0
+					for _, instance := range instances {
+						if instance.Vm.State == wantedState {
+							if instance.Vm.TemplateUUID == wantedInstanceTemplate {
+								var instanceTime time.Time
+								var err error
+								if instance.Vm.State != "Started" {
+									instanceTime, err = time.Parse(time.RFC3339, instance.Vm.LastUpdateTime) // can't use CreationTime because it's not updated for non-started instances
```

Comment:
@gjasny Please review

Reply:
The idea of the age was to see how old the oldest instance (of a certain template) is. If the timestamp that is used to calculate the age is now updated as well, the "birth date" of the instance is moving which defeats the purpose. E.g. in the case of pulling, the instance is now reported with a maximum value of 5 seconds.

If the `cr_time` field is not present for the `Scheduling` state (and others) we should exclude those states (or only include states which are guaranteed to have the `cr_time` field). Or skip all instances that don't have the `cr_time` field.

Reply:
On my controller even a "Scheduling" instance has the `cr_time` filed:
```json
      "vm": {
        "instance_id": "050823b7-ab5c-4947-77bb-06aa2a4a69d6",
        "instance_state": "Scheduling",
        "cr_time": "2025-01-30T07:55:07.334254871Z",
```

Reply:
@arturmelanchyk , can you check to see which states use `cr_time` and which use `ts`?

Reply:
in the most recent version of the controller (1.45.0) all instances in all states have both `cr_time` and `ts`

Reply:
> cr_time only set on an instance creation and never changed
> ts gets updated from time to time due to different events, like save image, termination etc
> both ts and cr_time are members of an Instance object, they do not depend on the vm (regardless if the vm has started or not)